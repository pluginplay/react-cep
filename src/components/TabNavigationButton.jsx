import styled from 'styled-components'

export default styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  background: ${props => props.isSelected ? props.theme.ae.panelBackgroundColor : props.theme.background.dark};
  cursor: pointer;
  transition: background 0.1s linear;
  width: 100%;
  height: 35px;
  border: none;
  outline: none;
  img {
    width: 16px;
    transition: transform 0.25s ease-in-out;
    transform: scale(1);
    margin: 0 auto;
  }
  &:hover {
    background: ${props => props.theme.ae.panelBackgroundColor};
  }
  &:active {
    padding: 0;
    img { transform: scale(0.9); }
  }
  @media (max-width: 320px) {
    height: 20px;
    img { width: 10px; }
  }
`
