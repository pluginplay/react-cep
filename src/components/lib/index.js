import SystemState from '../../models/SystemState'
import PersistedState from '../../models/PersistedState'

export { SystemState, PersistedState }
