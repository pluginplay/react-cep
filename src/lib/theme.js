import * as csInterface from 'cep-lib/csinterface'
import { lighten, darken } from 'polished'

export const getDefaultTheme = () => {
  const ae = csInterface.getThemeInformation()
  ae.systemHighlightColor = '#0078d7' // We have to hard-code this because Windows gets the wrong color.
  return {
    ae,
    background: {
      lighter: lighten(0.2, ae.panelBackgroundColor),
      light: lighten(0.1, ae.panelBackgroundColor),
      dark: darken(0.05, ae.panelBackgroundColor),
      darker: darken(0.1, ae.panelBackgroundColor)
    },
    colors: {
      green: '#23EB87',
      red: '#eb093f',
      blue: ae.systemHighlightColor
    }
  }
}
