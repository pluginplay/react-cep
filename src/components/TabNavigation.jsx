import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import TabNavigationButton from './TabNavigationButton'

class TabNavigation extends React.Component {
  state = {
    tabClickEvents: {}
  }

  constructor (props) {
    super(props)
    this.state.tabClickEvents = this._getTabClickEvents(props.tabs)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.tabs !== this.props.tabs) {
      this.setState({ tabClickEvents: this._getTabClickEvents(nextProps.tabs) })
    }
  }

  _getTabClickEvents (tabs) {
    return tabs.reduce((events, tab) => {
      events[tab.identifier] = this.onTabClick.bind(this, tab.identifier)
      return events
    }, {})
  }

  onTabClick (identifier) {
    this.props.onTabSelected(identifier)
  }

  render () {
    const tabButtons = this.props.tabs.map((tab) => {
      const isSelected = this.props.selectedTab === tab.identifier
      const image = `icons/${tab.icon}.svg`
      return (
        <TabNavigationButton key={tab.identifier} isSelected={isSelected}
          onClick={this.state.tabClickEvents[tab.identifier]}>
          <img src={image} alt={tab.title} />
        </TabNavigationButton>
      )
    })
    return (
      <div className={this.props.className}>
        {tabButtons}
      </div>
    )
  }
}

TabNavigation.propTypes = {
  tabs: PropTypes.arrayOf(PropTypes.object).isRequired,
  selectedTab: PropTypes.string.isRequired,
  onTabSelected: PropTypes.func.isRequired
}

export default styled(TabNavigation)`
  margin: 0 0 0 auto;
  background: ${props => props.theme.background.dark};
  width: 35px;
`
