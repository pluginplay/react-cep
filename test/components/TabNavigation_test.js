import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { expect, withTheme } from '../setup'
import { getDefaultTheme } from '../../src/lib/theme'
import TabNavigation from '../../src/components/TabNavigation'

import TabNavigationButton from '../../src/components/TabNavigationButton'

let tabs = null

beforeEach(() => {
  tabs = [
    { identifier: 'tab-1', title: 'Tab 1', icon: 'one' },
    { identifier: 'tab-2', title: 'Tab 2', icon: 'two' },
    { identifier: 'tab-3', title: 'Tab 3', icon: 'three' }
  ]
})

const getWrapper = (props = {}) => {
  return shallow(<TabNavigation theme={getDefaultTheme()} tabs={tabs} selectedTab={'tab-2'} onTabSelected={() => null} {...props} />).dive()
}

it('renders correctly', () => {
  const tree = renderer
    .create(withTheme(<TabNavigation tabs={tabs} selectedTab={'tab-2'} onTabSelected={() => null} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})

it('renders the configured tabs', () => {
  const wrapper = getWrapper()
  const navigationButtons = wrapper.find(TabNavigationButton)
  expect(navigationButtons).to.have.length(3)
  expect(navigationButtons.at(0).key()).to.equal('tab-1')
  expect(navigationButtons.at(1).key()).to.equal('tab-2')
  expect(navigationButtons.at(2).key()).to.equal('tab-3')
  expect(navigationButtons.at(0).find('img').first()).to.have.prop('src').equal('icons/one.svg')
  expect(navigationButtons.at(1).find('img').first()).to.have.prop('src').equal('icons/two.svg')
  expect(navigationButtons.at(2).find('img').first()).to.have.prop('src').equal('icons/three.svg')
  expect(navigationButtons.at(0).find('img').first()).to.have.prop('alt').equal('Tab 1')
  expect(navigationButtons.at(1).find('img').first()).to.have.prop('alt').equal('Tab 2')
  expect(navigationButtons.at(2).find('img').first()).to.have.prop('alt').equal('Tab 3')
})

it('calls onTabSelected when a new tab is selected', () => {
  const onTabSelected = jest.fn()
  const wrapper = getWrapper({ onTabSelected })
  wrapper.find(TabNavigationButton).first().simulate('click')
  jestExpect(onTabSelected).toHaveBeenCalledTimes(1)
  jestExpect(onTabSelected).toHaveBeenCalledWith('tab-1')
})

it('marks selectedTab as selected', () => {
  const wrapper = getWrapper({ selectedTab: 'tab-1' })
  const firstTab = wrapper.find(TabNavigationButton).first()
  expect(firstTab).to.have.prop('isSelected').equal(true)
  expect(firstTab.key()).to.equal('tab-1')
})
