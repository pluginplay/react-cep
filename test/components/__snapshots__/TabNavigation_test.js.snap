// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bwzfXH cYfKZY"
>
  <button
    className="sc-bdVaJa gPiPUB"
    onClick={[Function]}
  >
    <img
      alt="Tab 1"
      src="icons/one.svg"
    />
  </button>
  <button
    className="sc-bdVaJa eazgxO"
    onClick={[Function]}
  >
    <img
      alt="Tab 2"
      src="icons/two.svg"
    />
  </button>
  <button
    className="sc-bdVaJa gPiPUB"
    onClick={[Function]}
  >
    <img
      alt="Tab 3"
      src="icons/three.svg"
    />
  </button>
</div>
`;
